Fix deleteByQuery fatal error

dimvic opened the pull request:
deleteByQuery was always throwing `Call to a member function toArray()  on a non-object` since $query->getQuery() returns array. Should probably also change Query->getQuery() phpdoc to typehint the return value as array.

ruflin commented:
Thanks for the fix. See also: https://github.com/ruflin/Elastica/pull/1037 Would you mind adding a test and updating CHANGELOG.md?

dimvic commented:
I modified it a little bit more so you can delete using both Query instances and array queries.

I also added a test, should be fine, though I am unable to run the tests:
```
Waiting on elasticsearch(elasticsearch:9200) to start................
Elasticsearch is not running or is not healthy.
Address: elasticsearch:9200

make: *** [run] Error 1
```
and Travis also fails for some reason irrelevant to the code itself. Could you look into that?

dimvic commented:
Well, I did update the changelog after all, along with the type hint for $query

ruflin commented:
Thanks for adding the test and updating the README. It seems like the test is currently failing:

https://travis-ci.org/ruflin/Elastica/jobs/111082687#L2178

```
1) Elastica\Test\IndexTest::testDeleteByQueryWithArrayQuery
Elastica\Exception\InvalidException: Param query does not exist
/elastica/lib/Elastica/Param.php:164
/elastica/lib/Elastica/Query.php:131
/elastica/lib/Elastica/Index.php:160
/elastica/test/lib/Elastica/Test/IndexTest.php:446
```

dimvic commented:
I fixed it, Travis keeps failing due to other kind of issues though

ruflin commented:
I restarted the failing build and hope it will pass now. Would you feel comfortable to squash the 5 commits into one?

ruflin commented:
@dimvic All green. Let me know if you feel comfortable with squashing and rebasing into one commit or if I should merge as it is.

dimvic commented:
squashed

ruflin commented:
Merged. Thanks a lot for this fix @dimvic

xjewer commented:
:tada:
